storage/provider/loop.go

```diff
@@ -4,87 +4,238 @@
 package provider
 
 import (
+	"fmt"
+	"os"
 	"path/filepath"
 	"strings"
 
 	"github.com/juju/errors"
+	"github.com/juju/names"
 
 	"github.com/juju/juju/environs/config"
 	"github.com/juju/juju/storage"
 )
 
 const (
-	// Loop provider types
+	// Loop provider types.
 	LoopProviderType     = storage.ProviderType("loop")
 	HostLoopProviderType = storage.ProviderType("hostloop")
-
-	// Config attributes
-	LoopDataDir = "data-dir" // top level directory where loop devices are created.
-	LoopSubDir  = "sub-dir"  // optional subdirectory for loop devices.
 )
 
 // loopProviders create volume sources which use loop devices.
-type loopProvider struct{}
+type loopProvider struct {
+	run runCommandFunc
+}
 
 var _ storage.Provider = (*loopProvider)(nil)
 
 // ValidateConfig is defined on the Provider interface.
-func (lp *loopProvider) ValidateConfig(providerConfig *storage.Config) error {
-	dataDir, ok := providerConfig.ValueString(LoopDataDir)
-	if !ok || dataDir == "" {
-		return errors.New("no data directory specified")
+func (lp *loopProvider) ValidateConfig(cfg *storage.Config) error {
+	// Loop provider has no configuration.
+	return nil
+}
+
+// validateFullConfig validates a fully-constructed storage config,
+// combining the user-specified config and any internally specified
+// config.
+func (lp *loopProvider) validateFullConfig(cfg *storage.Config) error {
+	if err := lp.ValidateConfig(cfg); err != nil {
+		return err
+	}
+	storageDir, ok := cfg.ValueString(storage.ConfigStorageDir)
+	if !ok || storageDir == "" {
+		return errors.New("storage directory not specified")
 	}
 	return nil
 }
 
 // VolumeSource is defined on the Provider interface.
-func (lp *loopProvider) VolumeSource(environConfig *config.Config, providerConfig *storage.Config) (storage.VolumeSource, error) {
-	if err := lp.ValidateConfig(providerConfig); err != nil {
+func (lp *loopProvider) VolumeSource(
+	environConfig *config.Config,
+	sourceConfig *storage.Config,
+) (storage.VolumeSource, error) {
+	if err := lp.validateFullConfig(sourceConfig); err != nil {
 		return nil, err
 	}
-	dataDir, _ := providerConfig.ValueString(LoopDataDir)
-	subDir, _ := providerConfig.ValueString(LoopDataDir)
-	return &loopVolumeSource{
-		dataDir,
-		subDir,
-	}, nil
+	storageDir, _ := sourceConfig.ValueString(storage.ConfigStorageDir)
```

Comment:
Given the possibility of coding errors, I think there should be a check that storageDir != ""